File: 2023/day-03/main2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func sumGearRatios(schematic []string) (int, error) {
	totalSum := 0
	symbolRegex := regexp.MustCompile(`[^0-9.]`) // Matches any non-digit and non-dot character

	for rowIdx, row := range schematic {
		for colIdx, char := range row {
			if symbolRegex.MatchString(string(char)) { // Found a symbol
				adjacentNumbers := []int{}

				for y := max(0, rowIdx-1); y <= min(len(schematic)-1, rowIdx+1); y++ {
					for x := max(0, colIdx-1); x <= min(len(row)-1, colIdx+1); x++ {
						if y == rowIdx && x == colIdx {
							continue // Skip the symbol itself
						}
						numStr := regexp.MustCompile(`\d+`).FindString(schematic[y][x:])
						if numStr != "" {
							num, err := strconv.Atoi(numStr)
							if err != nil {
								return 0, fmt.Errorf("error converting number '%s' to integer: %v", numStr, err)
							}
							adjacentNumbers = append(adjacentNumbers, num)
						}
					}
				}

				if len(adjacentNumbers) == 2 { // Gear must have exactly two adjacent part numbers
					gearRatio := adjacentNumbers[0] * adjacentNumbers[1]
					fmt.Printf("Found gear at position (%d, %d) with ratio: %d\n", rowIdx, colIdx, gearRatio)
					totalSum += gearRatio
				}
			}
		}
	}
	return totalSum, nil
}

func main() {
	// Read the engine schematic from a file
	file, err := os.Open("input.txt") // Replace with your file path
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var schematic []string
	for scanner.Scan() {
		schematic = append(schematic, scanner.Text())
	}

	gearRatioSum, err := sumGearRatios(schematic)
	if err != nil {
		fmt.Println("Error calculating gear ratio sum:", err)
		return
	}
	fmt.Println("The sum of all gear ratios in the engine schematic is:", gearRatioSum)
}

// max returns the maximum of two integers
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// min returns the minimum of two integers
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 2023/day-03/main2.py
import re

def sum_gear_ratios(schematic):
    total_sum = 0
    symbol_regex = re.compile(r"[^\d.]")

    for row_idx, row in enumerate(schematic):
        for col_idx, char in enumerate(row):
            if symbol_regex.match(char):
                adjacent_numbers = []
                for dy in [-1, 0, 1]:
                    for dx in [-1, 0, 1]:
                        new_row = row_idx + dy
                        new_col = col_idx + dx
                        if (
                                0 <= new_row < len(schematic) and
                                0 <= new_col < len(row) and
                                (new_row != row_idx or new_col != col_idx) # Exclude the symbol itself
                        ):
                            match = re.search(r"\d+", schematic[new_row][new_col:])
                            if match and match.start() == 0: # ensure the number starts at the current position
                                number = int(match.group())
                                adjacent_numbers.append(number)

                if len(adjacent_numbers) == 2:
                    gear_ratio = adjacent_numbers[0] * adjacent_numbers[1]
                    print(f"Found gear at position ({row_idx}, {col_idx}) with ratio: {gear_ratio}")
                    total_sum += gear_ratio

    return total_sum


# Read the schematic from the file
file_path = "input.txt"
with open(file_path, "r") as file:
    schematic = file.read().splitlines()

# Calculate the sum of gear ratios
gear_ratio_sum = sum_gear_ratios(schematic)
print("The sum of all gear ratios in the engine schematic is:", gear_ratio_sum)

